opencog/atoms/distvalue/DistributionalValueSCM.cc

```diff
@@ -0,0 +1,463 @@
+	                        &DistributionalValueSCM::ss_new_cdv,
+	                        this,"distvalue");
+	define_scheme_primitive("cog-cdv-get-conditions",
+	                        &DistributionalValueSCM::ss_cdv_get_conditions,
+	                        this,"distvalue");
+	define_scheme_primitive("cog-cdv-get-unconditionals",
+	                        &DistributionalValueSCM::ss_cdv_get_unconditonals,
+	                        this,"distvalue");
+	define_scheme_primitive("cog-cdv-get-unconditional",
+	                        &DistributionalValueSCM::ss_cdv_get_unconditonal,
+	                        this,"distvalue");
+	define_scheme_primitive("cog-cdv-get-joint",
+	                        &DistributionalValueSCM::ss_cdv_get_joint,
+	                        this,"distvalue");
+	define_scheme_primitive("cog-cdv-merge",
+	                        &DistributionalValueSCM::ss_cdv_merge,
+	                        this,"distvalue");
+	define_scheme_primitive("cog-cdv-cde",
+	                        &DistributionalValueSCM::ss_cdv_cde,
+	                        this,"distvalue");
+}
+
+//Utility Functions
+
+Interval
+DistributionalValueSCM::verify_interval(SCM svalue_list, const char * subrname, int pos)
+{
+	// Verify that second arg is an actual list.
+	// Null lists are not valid intervals.
+	if (!scm_is_pair(svalue_list))
+		scm_wrong_type_arg_msg(subrname, pos, svalue_list, "a non-null list of float-pt values");
+	return SchemeSmob::scm_to_float_list(svalue_list);
+}
+
+DVKey
+DistributionalValueSCM::verify_DVKey(SCM svalue_list, const char * subrname, int pos)
+{
+	// Verify that second arg is an actual list.
+	// Null lists are not valid intervals.
+	if (!scm_is_pair(svalue_list))
+		scm_wrong_type_arg_msg(subrname, pos, svalue_list, "a non-null list of float-pt values");
+	return scm_to_DVKey(svalue_list);
+}
+
+DVKeySeq
+DistributionalValueSCM::verify_DVKeySeq(SCM svalue_list, const char * subrname, int pos)
+{
+	// Verify that second arg is an actual list.
+	// Null lists are not valid intervals.
+	if (!scm_is_pair(svalue_list))
+		scm_wrong_type_arg_msg(subrname, pos, svalue_list, "a non-null list of float-pt values");
+	return scm_to_DVKeySeq(svalue_list);
+}
+
+DVKey DistributionalValueSCM::scm_to_DVKey(SCM svalue_list)
+{
+	DVKey valist;
+	SCM sl = svalue_list;
+	while (scm_is_pair(sl)) {
+		SCM svalue = SCM_CAR(sl);
+
+		if (not scm_is_null(svalue)) {
+			std::vector<double> v = verify_interval(svalue,"scm_to_DVKey",0);
+			valist.emplace_back(v);
+		}
+		sl = SCM_CDR(sl);
+	}
+	return valist;
+}
+
+DVKeySeq DistributionalValueSCM::scm_to_DVKeySeq(SCM svalue_list)
+{
+	DVKeySeq valist;
+	SCM sl = svalue_list;
+	while (scm_is_pair(sl)) {
+		SCM svalue = SCM_CAR(sl);
+
+		if (not scm_is_null(svalue)) {
+			DVKey v = verify_DVKey(svalue,"scm_to_DVKeySeq",0);
+			valist.emplace_back(v);
+		}
+		sl = SCM_CDR(sl);
+	}
+	return valist;
+}
+
+SCM DistributionalValueSCM::float_list_to_scm(const std::vector<double>& v)
+{
```

Comment:
FYI, there is a scheme function `cog-value->list` implemented in `ss_value_to_list` that accepts an SCM value, and returns a list of whatever is inside of it. It is currently implemented as one big case statement; it should probably be converted to some kind of per-value callback (so that each value knows how to convert itself into a list).  We've never done this kind of per-object scheme callback, so not clear how that might work.

Its plausible to think that someone someday might expect `cog-value->list` to do something reasonable for distributional values; however, implementing this, today, is low priority.